openedx/core/lib/request_utils.py

```diff
@@ -227,3 +229,251 @@ def set_custom_attributes_for_top_n(self, names_to_size, top_n_captured, attribu
+        log.exception(f'Error processing setting EXPECTED_ERRORS. {repr(e)}')
+
+    _EXPECTED_ERROR_SETTINGS_DICT = expected_error_settings_dict
+    return _EXPECTED_ERROR_SETTINGS_DICT
+
+
+def clear_cached_expected_error_settings():
+    """
+    Clears the cached expected error settings. Useful for testing.
+    """
+    global _EXPECTED_ERROR_SETTINGS_DICT
+    _EXPECTED_ERROR_SETTINGS_DICT = None
+
+
+def _log_and_monitor_expected_errors(request, exception, caller):
+    """
+    Adds logging and monitoring for expected errors as needed.
+
+    Arguments:
+        request: The request
+        exception: The exception
+        caller: Either 'middleware' or 'drf`
+    """
+    expected_error_settings_dict = _get_expected_error_settings_dict()
+    if not expected_error_settings_dict:
+        return
+
+    # 'module.Class', for example, 'django.core.exceptions.PermissionDenied'
+    # Note: `Exception` itself doesn't have a module.
+    exception_module = getattr(exception, '__module__', '')
+    separator = '.' if exception_module else ''
+    module_and_class = f'{exception_module}{separator}{exception.__class__.__name__}'
+
+    # Set checked_error_expected_from custom attribute to potentially help find issues where errors are never processed.
+    set_custom_attribute('checked_error_expected_from', caller)
+
+    # check if we already added logging/monitoring from a different caller
+    request_cache = RequestCache('openedx.core.lib.request_utils')
+    cached_handled_exception = request_cache.get_cached_response('handled_exception')
+    if cached_handled_exception.is_found:
+        cached_module_and_class = cached_handled_exception.value
+        # exception was already processed by a different caller
+        if cached_handled_exception.value == module_and_class:
+            set_custom_attribute('checked_error_expected_from', 'multiple')
+            return
+
+        # Currently, it seems unexpected that middleware and drf will both handle different uncaught exceptions.
+        # However, since it is possible, we will add an additional attribute and log message and then continue.
+        set_custom_attribute('unexpected_multiple_exceptions', cached_module_and_class)
+        log.warning(
+            "Unexpected scenario where different exceptions are handled by _log_and_monitor_expected_errors. "
+            "See 'unexpected_multiple_exceptions' custom attribute."
+        )
+    request_cache.set('handled_exception', module_and_class)
+
+    if module_and_class not in expected_error_settings_dict:
+        return
+
+    exception_message = repr(str(exception))
+    module_and_class_with_message = f'{module_and_class}({exception_message})'
+    set_custom_attribute('error_expected', module_and_class_with_message)
+
+    expected_error_settings = expected_error_settings_dict[module_and_class]
+    if expected_error_settings['is_ignored']:
+        set_custom_attribute('error_ignored', True)
```

Comment:
It occurs to me that if there _were_ instances of `unexpected_multiple_exceptions`, we could end up with inconsistent attributes -- if the first set `error_ignored=True`, and the second didn't set the attribute at all, the attribute would still stick around. If we have an alert on that attribute showing up, it's probably fine, but I wonder if we shouldn't maybe _always_ set the attribute. Probably not a big deal, but... thoughts?

Reply:
I added `unexpected_multiple_exceptions` because I'm not quite sure what we'll see. I plan to add an alert for this, and we can determine how and if we deal with it then.  I might add a light runbook with your note as well.  I don't want to over invest in case this never can happen.

Reply:
Yup, sounds fine.

Reply:
Good reminder though.  Thanks.  I'll add a note to my alerting subtask.

Reply:
Thinking about it more... I think I have a strong enough preference for consistency that even though this is an unlikely scenario, it would be worth always sending the attribute. (If weird stuff is happening, I'd like it to have as small a blast radius as possible.) This could be changed to `set_custom_attribute('error_ignored', expected_error_settings['is_ignored'])` without too much trouble, right?

Reply:
(If this required a bunch of extra conditionals, though, I'd say skip it.)

Reply:
Right now, `error_ignored` is always True and is meant to be used with `WHERE error_ignored IS NULL`, or `WHERE error_ignored IS NOT NULL`. I'd rather not change values.

Instead, I updated this code to just print a warning with the new `module.Class` and return without processing the second exception.  Make sense?

Reply:
That works too!

Reply:
I also just added a comment about the expected querying of it in the  toggle annotation.